File: api-go/internal/services/drive_upload.go
package services

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"time"

	"github.com/cymoo/mote/internal/config"
	"github.com/cymoo/mote/internal/models"
	"github.com/jmoiron/sqlx"
)

const (
	driveUploadTTL    = 24 * time.Hour
	maxFileSize       = int64(4 << 30) // 4 GB
	defaultChunkSize  = int64(8 << 20) // 8 MB
)

var (
	ErrUploadNotFound       = errors.New("upload session not found or expired")
	ErrUploadChunkSize      = errors.New("chunk size mismatch")
	ErrUploadChunkOOR       = errors.New("chunk index out of range")
	ErrUploadIncomplete     = errors.New("upload is incomplete")
	ErrUploadFinalSize      = errors.New("final file size mismatch")
	ErrUploadTooLarge       = errors.New("file too large")
	ErrUploadInvalidRequest = errors.New("invalid upload request")
)

// DriveUploadService handles chunked, resumable uploads.
type DriveUploadService struct {
	db     *sqlx.DB
	drive  *DriveService
	config *config.UploadConfig
}

func NewDriveUploadService(db *sqlx.DB, drive *DriveService, cfg *config.UploadConfig) *DriveUploadService {
	return &DriveUploadService{db: db, drive: drive, config: cfg}
}

func (s *DriveUploadService) chunksDir(uploadID string) string {
	return filepath.Join(s.config.BasePath, "drive", "_chunks", uploadID)
}

// Init creates a new upload session.
func (s *DriveUploadService) Init(ctx context.Context, req models.DriveUploadInitRequest) (*models.DriveUpload, error) {
	if err := validName(req.Name); err != nil {
		return nil, err
	}
	if req.Size <= 0 || req.Size > maxFileSize {
		return nil, ErrUploadTooLarge
	}
	chunk := req.ChunkSize
	if chunk <= 0 {
		chunk = defaultChunkSize
	}
	if chunk < 1<<20 || chunk > 64<<20 {
		return nil, ErrUploadInvalidRequest
	}
	if req.ParentID != nil {
		if _, err := s.drive.requireActiveFolder(ctx, *req.ParentID); err != nil {
			return nil, err
		}
	}

	total := int((req.Size + chunk - 1) / chunk)
	mask := make([]byte, (total+7)/8)

	id := newToken(16)
	now := time.Now().UnixMilli()
	expiresAt := now + driveUploadTTL.Milliseconds()

	if err := os.MkdirAll(s.chunksDir(id), 0755); err != nil {
		return nil, err
	}

	_, err := s.db.ExecContext(ctx, `
INSERT INTO drive_uploads (id, parent_id, name, size, chunk_size, total_chunks, received_mask, status, expires_at, created_at, updated_at)
VALUES (?, ?, ?, ?, ?, ?, ?, 'uploading', ?, ?, ?)`,
		id, req.ParentID, req.Name, req.Size, chunk, total, mask, expiresAt, now, now)
	if err != nil {
		_ = os.RemoveAll(s.chunksDir(id))
		return nil, err
	}

	return s.find(ctx, id)
}

func (s *DriveUploadService) find(ctx context.Context, id string) (*models.DriveUpload, error) {
	var u models.DriveUpload
	err := s.db.GetContext(ctx, &u, `SELECT * FROM drive_uploads WHERE id = ?`, id)
	if err == sql.ErrNoRows {
		return nil, ErrUploadNotFound
	}
	if err != nil {
		return nil, err
	}
	if u.ExpiresAt < time.Now().UnixMilli() {
		return nil, ErrUploadNotFound
	}
	return &u, nil
}

// Get returns a session and the list of received chunk indices.
func (s *DriveUploadService) Get(ctx context.Context, id string) (*models.DriveUpload, []int, error) {
	u, err := s.find(ctx, id)
	if err != nil {
		return nil, nil, err
	}
	return u, decodeMask(u.ReceivedMask, u.TotalChunks), nil
}

// PutChunk writes a single chunk idempotently.
func (s *DriveUploadService) PutChunk(ctx context.Context, id string, idx int, body io.Reader) error {
	u, err := s.find(ctx, id)
	if err != nil {
		return err
	}
	if u.Status != "uploading" {
		return ErrUploadInvalidRequest
	}
	if idx < 0 || idx >= u.TotalChunks {
		return ErrUploadChunkOOR
	}

	expected := u.ChunkSize
	if idx == u.TotalChunks-1 {
		// last chunk may be smaller
		expected = u.Size - int64(idx)*u.ChunkSize
	}

	tmp := filepath.Join(s.chunksDir(id), fmt.Sprintf("%d.part", idx))
	final := filepath.Join(s.chunksDir(id), fmt.Sprintf("%d.bin", idx))

	f, err := os.Create(tmp)
	if err != nil {
		return err
	}
	written, err := io.CopyN(f, body, expected+1) // read one extra to detect oversize
	if err != nil && err != io.EOF {
		f.Close()
		_ = os.Remove(tmp)
		return err
	}
	if err := f.Close(); err != nil {
		_ = os.Remove(tmp)
		return err
	}
	if written != expected {
		_ = os.Remove(tmp)
		return ErrUploadChunkSize
	}
	if err := os.Rename(tmp, final); err != nil {
		_ = os.Remove(tmp)
		return err
	}

	// Update bitmap atomically. We use BEGIN IMMEDIATE to acquire the
	// reserved lock up-front: a deferred transaction (the default) starts
	// with a SHARED lock on SELECT and then upgrades on UPDATE, which can
	// deadlock when multiple chunk uploads race for the same row and one
	// returns SQLITE_BUSY despite busy_timeout.
	conn, err := s.db.Connx(ctx)
	if err != nil {
		return err
	}
	defer conn.Close()
	if _, err := conn.ExecContext(ctx, "BEGIN IMMEDIATE"); err != nil {
		return err
	}
	committed := false
	defer func() {
		if !committed {
			_, _ = conn.ExecContext(context.Background(), "ROLLBACK")
		}
	}()
	var mask []byte
	if err := conn.GetContext(ctx, &mask, `SELECT received_mask FROM drive_uploads WHERE id = ?`, id); err != nil {
		return err
	}
	if idx/8 >= len(mask) {
		return ErrUploadInvalidRequest
	}
	mask[idx/8] |= 1 << (idx % 8)
	now := time.Now().UnixMilli()
	if _, err := conn.ExecContext(ctx,
		`UPDATE drive_uploads SET received_mask = ?, updated_at = ? WHERE id = ?`,
		mask, now, id); err != nil {
		return err
	}
	if _, err := conn.ExecContext(ctx, "COMMIT"); err != nil {
		return err
	}
	committed = true
	return nil
}

// Complete assembles all chunks into the final blob and inserts a drive_nodes row.
// onCollision: "ask" (default; returns ErrDriveNameConflict on conflict),
// "overwrite", "rename", or "skip".
// On "skip" with collision, returns the existing node and discards the upload.
func (s *DriveUploadService) Complete(
	ctx context.Context, id string, onCollision string,
) (*models.DriveNode, error) {
	u, err := s.find(ctx, id)
	if err != nil {
		return nil, err
	}
	// Only one assemble at a time per session.
	res, err := s.db.ExecContext(ctx,
		`UPDATE drive_uploads SET status = 'assembling', updated_at = ?
		 WHERE id = ? AND status = 'uploading'`,
		time.Now().UnixMilli(), id)
	if err != nil {
		return nil, err
	}
	if n, _ := res.RowsAffected(); n == 0 {
		return nil, ErrUploadInvalidRequest
	}
	defer func() {
		// On error, allow the client to retry.
		if r := recover(); r != nil {
			if _, err := s.db.ExecContext(context.Background(),
				`UPDATE drive_uploads SET status = 'uploading' WHERE id = ?`, id); err != nil {
				log.Printf("drive upload: revert status after panic failed: %v", err)
			}
			panic(r)
		}
	}()

	// Verify all chunks present.
	var mask []byte
	if err := s.db.GetContext(ctx, &mask, `SELECT received_mask FROM drive_uploads WHERE id = ?`, id); err != nil {
		s.markUploading(id)
		return nil, err
	}
	if len(mask)*8 < u.TotalChunks {
		s.markUploading(id)
		return nil, ErrUploadInvalidRequest
	}
	for i := 0; i < u.TotalChunks; i++ {
		if mask[i/8]&(1<<(i%8)) == 0 {
			s.markUploading(id)
			return nil, ErrUploadIncomplete
		}
	}

	// Pre-collision check (fast path for skip/rename without writing the blob to drive/).
	var parentID *int64
	if u.ParentID.Valid {
		v := u.ParentID.Int64
		parentID = &v
	}
	// Re-validate parent: it may have been (soft-)deleted between Init and now.
	if parentID != nil {
		if _, err := s.drive.requireActiveFolder(ctx, *parentID); err != nil {
			s.markUploading(id)
			return nil, err
		}
	}
	finalName := u.Name
	if existing, err := s.drive.FindActiveSibling(ctx, parentID, finalName); err != nil {
		s.markUploading(id)
		return nil, err
	} else if existing != nil {
		switch onCollision {
		case "skip":
			_ = s.deleteSession(id)
			return existing, nil
		case "rename":
			finalName, err = s.drive.AutoRename(ctx, parentID, finalName)
			if err != nil {
				s.markUploading(id)
				return nil, err
			}
		case "overwrite":
			// fall through; ReplaceFileNode handles it.
		default: // "ask" or empty
			s.markUploading(id)
			return nil, ErrDriveNameConflict
		}
	}

	// Assemble into a temp file inside drive/, then rename.
	blobName := newBlobName(finalName)
	relPath := filepath.Join("drive", blobName)
	absPath := filepath.Join(s.config.BasePath, relPath)
	tmpAbs := absPath + ".part"

	hash, written, err := s.assemble(u, tmpAbs)
	if err != nil {
		_ = os.Remove(tmpAbs)
		s.markUploading(id)
		return nil, err
	}
	if written != u.Size {
		_ = os.Remove(tmpAbs)
		s.markUploading(id)
		return nil, ErrUploadFinalSize
	}
	if err := os.Rename(tmpAbs, absPath); err != nil {
		_ = os.Remove(tmpAbs)
		s.markUploading(id)
		return nil, err
	}

	var node *models.DriveNode
	if onCollision == "overwrite" {
		node, err = s.drive.ReplaceFileNode(ctx, parentID, finalName, relPath, hash, u.Size)
	} else {
		node, err = s.drive.CreateFileNode(ctx, parentID, finalName, relPath, hash, u.Size)
	}
	if err != nil {
		_ = os.Remove(absPath)
		s.markUploading(id)
		return nil, err
	}

	_ = s.deleteSession(id)
	return node, nil
}

// Cancel aborts a session and removes its chunks. It only operates on
// 'uploading' sessions to avoid racing the assemble phase of Complete.
func (s *DriveUploadService) Cancel(ctx context.Context, id string) error {
	res, err := s.db.ExecContext(ctx,
		`DELETE FROM drive_uploads WHERE id = ? AND status = 'uploading'`, id)
	if err != nil {
		return err
	}
	if n, _ := res.RowsAffected(); n == 0 {
		// Either gone already, or in 'assembling' — leave Complete to finish.
		return nil
	}
	return os.RemoveAll(s.chunksDir(id))
}

// PurgeExpired removes upload sessions past their expires_at.
func (s *DriveUploadService) PurgeExpired(ctx context.Context) (int, error) {
	now := time.Now().UnixMilli()
	var ids []string
	if err := s.db.SelectContext(ctx, &ids,
		`SELECT id FROM drive_uploads WHERE expires_at < ?`, now); err != nil {
		return 0, err
	}
	for _, id := range ids {
		_ = s.deleteSession(id)
	}
	return len(ids), nil
}

func (s *DriveUploadService) deleteSession(id string) error {
	_, _ = s.db.Exec(`DELETE FROM drive_uploads WHERE id = ?`, id)
	return os.RemoveAll(s.chunksDir(id))
}

func (s *DriveUploadService) markUploading(id string) {
	s.db.Exec(`UPDATE drive_uploads SET status = 'uploading' WHERE id = ?`, id)
}

// assemble concatenates chunk files into outPath while computing sha256.
func (s *DriveUploadService) assemble(u *models.DriveUpload, outPath string) (string, int64, error) {
	out, err := os.Create(outPath)
	if err != nil {
		return "", 0, err
	}
	defer out.Close()

	h := sha256.New()
	mw := io.MultiWriter(out, h)

	var total int64
	for i := 0; i < u.TotalChunks; i++ {
		path := filepath.Join(s.chunksDir(u.ID), fmt.Sprintf("%d.bin", i))
		f, err := os.Open(path)
		if err != nil {
			return "", 0, err
		}
		n, err := io.Copy(mw, f)
		f.Close()
		if err != nil {
			return "", 0, err
		}
		total += n
	}
	if err := out.Sync(); err != nil {
		return "", 0, err
	}
	return hex.EncodeToString(h.Sum(nil)), total, nil
}

func decodeMask(mask []byte, total int) []int {
	out := make([]int, 0, total)
	for i := 0; i < total; i++ {
		if i/8 >= len(mask) {
			break
		}
		if mask[i/8]&(1<<(i%8)) != 0 {
			out = append(out, i)
		}
	}
	return out
}

File: api-py/app/services/drive_upload.py
"""Drive upload service — chunked uploads.

Translation of api-go/internal/services/drive_upload.go.
"""
from __future__ import annotations

import hashlib
import os
import shutil
import threading
import uuid
from dataclasses import dataclass
from typing import Optional

from sqlalchemy import text
from sqlalchemy.exc import IntegrityError

from ..extension import db
from ..model import utc_now_ms
from .drive import (
    DriveError,
    DriveNameConflict,
    DriveNotFound,
    DriveService,
    new_blob_name,
    valid_name,
)


MAX_FILE_SIZE = 4 << 30  # 4 GiB
DEFAULT_CHUNK_SIZE = 8 << 20  # 8 MiB
MIN_CHUNK_SIZE = 1 << 20  # 1 MiB
MAX_CHUNK_SIZE = 64 << 20  # 64 MiB
UPLOAD_TTL_MS = 24 * 60 * 60 * 1000


class UploadNotFound(DriveError):
    pass


class UploadInvalid(DriveError):
    pass


class UploadCollision(DriveError):
    """Raised when on_collision='ask' and a same-name file already exists."""


class UploadGone(DriveError):
    pass


@dataclass
class UploadStatus:
    upload_id: str
    parent_id: Optional[int]
    name: str
    size: int
    chunk_size: int
    total_chunks: int
    received_chunks: list[int]
    status: str

    def to_dict(self) -> dict:
        return {
            'upload_id': self.upload_id,
            'parent_id': self.parent_id,
            'name': self.name,
            'size': self.size,
            'chunk_size': self.chunk_size,
            'total_chunks': self.total_chunks,
            'received_chunks': self.received_chunks,
            'status': self.status,
        }


def _decode_mask(buf: bytes, total: int) -> list[int]:
    out = []
    for i in range(total):
        if buf[i // 8] & (1 << (i % 8)):
            out.append(i)
    return out


class DriveUploadService:
    # Per-upload locks so concurrent PUT chunks for the same upload don't
    # race when OR-masking the bitmap. Cross-process safety relies on SQLite
    # transactions; this just removes intra-process contention.
    _locks: dict[str, threading.Lock] = {}
    _locks_guard = threading.Lock()

    def __init__(self, drive: DriveService):
        self.drive = drive

    # -- helpers --------------------------------------------------------------

    @classmethod
    def _lock_for(cls, upload_id: str) -> threading.Lock:
        with cls._locks_guard:
            l = cls._locks.get(upload_id)
            if l is None:
                l = threading.Lock()
                cls._locks[upload_id] = l
        return l

    def _chunks_dir(self, upload_id: str) -> str:
        return os.path.join(self.drive.base_path, 'drive', '_chunks', upload_id)

    def _chunk_path(self, upload_id: str, idx: int) -> str:
        return os.path.join(self._chunks_dir(upload_id), f'{idx:06d}')

    # -- API ------------------------------------------------------------------

    def init(
        self,
        parent_id: Optional[int],
        name: str,
        size: int,
        chunk_size: Optional[int],
    ) -> UploadStatus:
        valid_name(name)
        if size <= 0 or size > MAX_FILE_SIZE:
            raise UploadInvalid('invalid size')
        cs = chunk_size or DEFAULT_CHUNK_SIZE
        if cs < MIN_CHUNK_SIZE or cs > MAX_CHUNK_SIZE:
            raise UploadInvalid('invalid chunk_size')
        if parent_id is not None:
            self.drive._require_active_folder(parent_id)
        total = (size + cs - 1) // cs
        mask = bytes((total + 7) // 8)
        upload_id = uuid.uuid4().hex
        now = utc_now_ms()
        try:
            db.session.execute(
                text(
                    'INSERT INTO drive_uploads '
                    '(id, parent_id, name, size, chunk_size, total_chunks, '
                    'received_mask, status, expires_at, created_at, updated_at) '
                    "VALUES (:id, :pid, :name, :size, :cs, :total, :mask, "
                    "'uploading', :exp, :now, :now)"
                ),
                {
                    'id': upload_id,
                    'pid': parent_id,
                    'name': name,
                    'size': size,
                    'cs': cs,
                    'total': total,
                    'mask': mask,
                    'exp': now + UPLOAD_TTL_MS,
                    'now': now,
                },
            )
            db.session.commit()
        except Exception:
            db.session.rollback()
            raise
        os.makedirs(self._chunks_dir(upload_id), exist_ok=True)
        return UploadStatus(
            upload_id=upload_id,
            parent_id=parent_id,
            name=name,
            size=size,
            chunk_size=cs,
            total_chunks=total,
            received_chunks=[],
            status='uploading',
        )

    def get(self, upload_id: str) -> UploadStatus:
        row = db.session.execute(
            text('SELECT * FROM drive_uploads WHERE id = :id'),
            {'id': upload_id},
        ).first()
        if row is None:
            raise UploadNotFound('upload not found')
        if row.expires_at < utc_now_ms() and row.status == 'uploading':
            raise UploadGone('upload session expired')
        return UploadStatus(
            upload_id=row.id,
            parent_id=row.parent_id,
            name=row.name,
            size=row.size,
            chunk_size=row.chunk_size,
            total_chunks=row.total_chunks,
            received_chunks=_decode_mask(row.received_mask, row.total_chunks),
            status=row.status,
        )

    def put_chunk(self, upload_id: str, idx: int, data: bytes) -> None:
        with self._lock_for(upload_id):
            row = db.session.execute(
                text('SELECT * FROM drive_uploads WHERE id = :id'),
                {'id': upload_id},
            ).first()
            if row is None:
                raise UploadNotFound('upload not found')
            if row.status != 'uploading':
                raise UploadInvalid(f'upload status is {row.status}')
            if row.expires_at < utc_now_ms():
                raise UploadGone('upload session expired')
            if idx < 0 or idx >= row.total_chunks:
                raise UploadInvalid('chunk index out of range')

            expected = row.chunk_size
            if idx == row.total_chunks - 1:
                expected = row.size - row.chunk_size * (row.total_chunks - 1)
            if len(data) != expected:
                raise UploadInvalid(
                    f'chunk size mismatch: got {len(data)} expected {expected}'
                )

            os.makedirs(self._chunks_dir(upload_id), exist_ok=True)
            tmp = self._chunk_path(upload_id, idx) + '.part'
            with open(tmp, 'wb') as f:
                f.write(data)
            os.replace(tmp, self._chunk_path(upload_id, idx))

            mask = bytearray(row.received_mask)
            mask[idx // 8] |= 1 << (idx % 8)
            try:
                db.session.execute(
                    text(
                        'UPDATE drive_uploads SET received_mask = :m, updated_at = :now '
                        "WHERE id = :id AND status = 'uploading'"
                    ),
                    {'m': bytes(mask), 'now': utc_now_ms(), 'id': upload_id},
                )
                db.session.commit()
            except Exception:
                db.session.rollback()
                raise

    def complete(self, upload_id: str, on_collision: str = 'ask') -> 'object':
        # Atomic uploading -> assembling flip
        try:
            res = db.session.execute(
                text(
                    "UPDATE drive_uploads SET status = 'assembling', updated_at = :now "
                    "WHERE id = :id AND status = 'uploading'"
                ),
                {'now': utc_now_ms(), 'id': upload_id},
            )
            db.session.commit()
        except Exception:
            db.session.rollback()
            raise
        if res.rowcount == 0:
            row = db.session.execute(
                text('SELECT status FROM drive_uploads WHERE id = :id'),
                {'id': upload_id},
            ).first()
            if row is None:
                raise UploadNotFound('upload not found')
            raise UploadInvalid(f'upload not in uploading state (was {row.status})')

        try:
            return self._assemble(upload_id, on_collision)
        except Exception:
            try:
                db.session.execute(
                    text(
                        "UPDATE drive_uploads SET status = 'failed', updated_at = :now "
                        'WHERE id = :id'
                    ),
                    {'now': utc_now_ms(), 'id': upload_id},
                )
                db.session.commit()
            except Exception:
                db.session.rollback()
            raise

    def _assemble(self, upload_id: str, on_collision: str):
        row = db.session.execute(
            text('SELECT * FROM drive_uploads WHERE id = :id'),
            {'id': upload_id},
        ).first()
        if row is None:
            raise UploadNotFound('upload not found')

        # Verify all chunks present.
        for i in range(row.total_chunks):
            if not (row.received_mask[i // 8] & (1 << (i % 8))):
                raise UploadInvalid(f'missing chunk {i}')

        name = row.name
        parent_id = row.parent_id

        # Pre-collision check.
        sib = self.drive.find_active_sibling(parent_id, name)
        if sib is not None:
            if on_collision == 'skip':
                # Discard upload; return existing.
                self._cleanup_upload(upload_id)
                return sib
            if on_collision == 'rename':
                name = self.drive.auto_rename(parent_id, name)
            elif on_collision == 'overwrite':
                pass
            else:
                raise UploadCollision('name already exists in this folder')

        blob_name = new_blob_name(name)
        blob_rel = os.path.join('drive', blob_name)
        blob_abs = self.drive.blob_abs_path(blob_rel)

        os.makedirs(os.path.dirname(blob_abs), exist_ok=True)
        h = hashlib.sha256()
        tmp = blob_abs + '.part'
        try:
            with open(tmp, 'wb') as out:
                for i in range(row.total_chunks):
                    p = self._chunk_path(upload_id, i)
                    with open(p, 'rb') as fr:
                        while True:
                            buf = fr.read(64 * 1024)
                            if not buf:
                                break
                            out.write(buf)
                            h.update(buf)
            os.replace(tmp, blob_abs)
        except Exception:
            try:
                os.remove(tmp)
            except OSError:
                pass
            raise

        try:
            if sib is not None and on_collision == 'overwrite':
                node = self.drive.replace_file_node(
                    parent_id, name, blob_rel, h.hexdigest(), row.size
                )
            else:
                node = self.drive.create_file_node(
                    parent_id, name, blob_rel, h.hexdigest(), row.size
                )
        except DriveNameConflict:
            try:
                os.remove(blob_abs)
            except OSError:
                pass
            raise

        # Mark done + cleanup chunks.
        try:
            db.session.execute(
                text(
                    "UPDATE drive_uploads SET status = 'done', updated_at = :now "
                    'WHERE id = :id'
                ),
                {'now': utc_now_ms(), 'id': upload_id},
            )
            db.session.commit()
        except Exception:
            db.session.rollback()
        shutil.rmtree(self._chunks_dir(upload_id), ignore_errors=True)
        return node

    def cancel(self, upload_id: str) -> None:
        try:
            db.session.execute(
                text('DELETE FROM drive_uploads WHERE id = :id'),
                {'id': upload_id},
            )
            db.session.commit()
        except Exception:
            db.session.rollback()
            raise
        shutil.rmtree(self._chunks_dir(upload_id), ignore_errors=True)

    def _cleanup_upload(self, upload_id: str) -> None:
        try:
            db.session.execute(
                text('DELETE FROM drive_uploads WHERE id = :id'),
                {'id': upload_id},
            )
            db.session.commit()
        except Exception:
            db.session.rollback()
        shutil.rmtree(self._chunks_dir(upload_id), ignore_errors=True)

    # -- background -----------------------------------------------------------

    def purge_expired(self) -> int:
        rows = db.session.execute(
            text(
                'SELECT id FROM drive_uploads '
                'WHERE expires_at < :now '
                "  AND status IN ('uploading', 'failed')"
            ),
            {'now': utc_now_ms()},
        ).all()
        ids = [r.id for r in rows]
        for uid in ids:
            self._cleanup_upload(uid)
        return len(ids)
